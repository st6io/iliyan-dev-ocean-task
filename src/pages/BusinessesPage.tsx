import { useEffect, useRef, useState } from 'react';

import { x } from '@xstyled/styled-components';
import { createGlobalStyle } from '@xstyled/styled-components';

import Layout from '@components/Layout';
import Table from '@components/Table';

import businessesData from './businesses.mock.json';

const GlobalStyle = createGlobalStyle`
  body {
    overflow: hidden;
  }
`;

const headers = ['name', 'description'];
const rows = businessesData.map(({ id, name, description }) => ({
  id,
  cells: [name, description],
}));

const BusinessesPage = () => {
  const [height, setHeight] = useState('100vh');
  const ref = useRef<any>();

  useEffect(() => {
    const onResize = () => {
      if (ref.current) {
        setHeight(`${window.innerHeight - ref.current.offsetTop}px`);
      }
    };

    onResize();

    window.addEventListener('resize', onResize);

    return () => window.removeEventListener('resize', onResize);
  }, []);

  return (
    <Layout>
      <GlobalStyle />

      <x.div ref={ref} h={height} overflow="auto">
        <Table variant="primary" headers={headers} rows={rows} w={1} />
      </x.div>
    </Layout>
  );
};

export default BusinessesPage;
