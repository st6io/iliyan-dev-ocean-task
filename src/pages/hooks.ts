import { gql, useApolloClient, useQuery } from '@apollo/client';

const BUSINESS_FRAGMENT = gql`
  fragment BusinessFragment on Business {
    id
    name
    description
    phone
    image
    email

    address {
      number
      street
      zip
      city
      country
    }
  }
`;

export const BUSINESSES_QUERY = gql`
  query Businesses {
    businesses @rest(type: "Business", path: "b/6231abada703bb67492d2b8f") {
      ...BusinessFragment
    }
  }
  ${BUSINESS_FRAGMENT}
`;

export const useBusinessesQuery = () => useQuery<{ businesses: Business[] }>(BUSINESSES_QUERY);

export const useBusinessQuery = (id?: string) => {
  // ensure businesses are fetched if not present in the cache
  const { loading, error } = useBusinessesQuery();

  const business: Business | null = useApolloClient().readFragment({
    id: `Business:${id}`,
    fragment: BUSINESS_FRAGMENT,
  });

  return {
    loading,
    error,
    data: { business },
  };
};
