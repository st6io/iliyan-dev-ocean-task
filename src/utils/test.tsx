import { ReactElement, ReactNode } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { render as nativeRender } from '@testing-library/react';

const render = (ui: ReactElement, { path = '/', ...options } = {}) => {
  const RouterWrapper = ({ children }: { children: ReactNode }) => (
    <MemoryRouter initialEntries={[path]}>{children}</MemoryRouter>
  );

  return nativeRender(ui, {
    wrapper: RouterWrapper,
    ...options,
  });
};

export * from '@testing-library/react';
export { render };
