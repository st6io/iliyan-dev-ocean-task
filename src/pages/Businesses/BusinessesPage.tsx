import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { useNavigate } from 'react-router-dom';

import { x } from '@xstyled/styled-components';
import { createGlobalStyle } from '@xstyled/styled-components';

import Error from '@components/Error';
import Layout from '@components/Layout';
import Table from '@components/Table';

import { useBusinessesQuery } from '../hooks';

const GlobalStyle = createGlobalStyle`
  body {
    overflow: hidden;
  }
`;

const headers = ['name', 'description'];

const BusinessesPage = () => {
  const navigate = useNavigate();
  const onRowClick = useCallback(({ id }: { id: string }) => navigate(`./${id}`), [navigate]);

  const { loading, error, data } = useBusinessesQuery();
  const businessesRows = useMemo(
    () =>
      data?.businesses
        ? data.businesses.map(({ id, name, description }) => ({
            id,
            cells: [name, description],
          }))
        : [],
    [data],
  );

  const [height, setHeight] = useState('100vh');
  const ref = useRef<any>();

  useEffect(() => {
    const onResize = () => {
      if (ref.current) {
        setHeight(`${window.innerHeight - ref.current.offsetTop}px`);
      }
    };

    onResize();

    window.addEventListener('resize', onResize);

    return () => window.removeEventListener('resize', onResize);
  }, []);

  return (
    <Layout>
      <GlobalStyle />

      {error ? (
        <Error />
      ) : loading ? (
        <x.div>Loading...</x.div>
      ) : (
        <x.div ref={ref} h={height} overflow="auto">
          <Table
            variant="primary"
            headers={headers}
            rows={businessesRows}
            onRowClick={onRowClick}
            w={1}
          />
        </x.div>
      )}
    </Layout>
  );
};

export default BusinessesPage;
