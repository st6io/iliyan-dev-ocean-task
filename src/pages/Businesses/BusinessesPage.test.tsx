import { Route, Routes, useParams } from 'react-router-dom';

import { InMemoryCache } from '@apollo/client';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';

import { fireEvent, render, screen, waitFor } from '@src/utils/test';

import BusinessesPage from './BusinessesPage';

import businessesMock from '../businesses.mock.json';
import { BUSINESSES_QUERY } from '../hooks';

const successQueryMock: MockedResponse = {
  request: {
    query: BUSINESSES_QUERY,
  },
  result: {
    data: {
      businesses: businessesMock,
    },
  },
};

const DummyBusinessPage = () => {
  const { businessId } = useParams();

  return <div>Business Page: {businessId}</div>;
};

const renderPage = (mock = successQueryMock) => ({
  ...render(
    <MockedProvider mocks={[mock]} cache={new InMemoryCache()}>
      <Routes>
        <Route path="businesses" element={<BusinessesPage />} />
        <Route path="businesses/:businessId" element={<DummyBusinessPage />} />
      </Routes>
    </MockedProvider>,
    {
      path: '/businesses',
    },
  ),

  queryTableContainer: () => screen.queryByRole('table')?.parentElement,
});

const waitToLoadData = () =>
  waitFor(() => expect(screen.queryByText('Loading...')).not.toBeInTheDocument());

describe('BusinessesPage', () => {
  it('should match snapshot', async () => {
    const { container } = renderPage();

    await waitToLoadData();

    expect(container.firstChild).toMatchSnapshot();
  });

  it.only('should navigate to business page', async () => {
    renderPage();

    await waitToLoadData();

    const [businessToClickOn] = businessesMock;

    fireEvent.click(screen.getByText(businessToClickOn.name));

    expect(screen.getByText(`Business Page: ${businessToClickOn.id}`)).toBeInTheDocument();
  });

  it('should wrap table in scrollable container', async () => {
    const { queryTableContainer } = renderPage();

    await waitToLoadData();

    expect(queryTableContainer()).toBeInTheDocument();
    expect(queryTableContainer()).toHaveStyle({ overflow: 'auto' });
  });

  it('should update table container height on window resize', async () => {
    const { queryTableContainer } = renderPage();
    const initialWindowHeight = window.innerHeight;
    const heightChange = -168;

    await waitToLoadData();

    expect(queryTableContainer()).toHaveStyle({
      height: '100vh',
    });

    window.innerHeight = initialWindowHeight + heightChange;
    fireEvent.resize(window);

    expect(queryTableContainer()).toHaveStyle({
      height: `${initialWindowHeight + heightChange}px`,
    });
  });

  it('should render loading screen', () => {
    renderPage();

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('should render error screen', async () => {
    const errorQueryMock = {
      request: {
        query: BUSINESSES_QUERY,
      },
      result: {
        data: undefined,
      },
      error: new Error('Something went wrong'),
    };
    renderPage(errorQueryMock);

    await waitToLoadData();

    expect(screen.getByText('Something went wrong.')).toBeInTheDocument();
    expect(screen.getByText('Please try again later.')).toBeInTheDocument();
  });
});
