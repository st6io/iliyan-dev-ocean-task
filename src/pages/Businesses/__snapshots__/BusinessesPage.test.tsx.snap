// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessesPage should match snapshot 1`] = `
<div
  class="sc-cxabCf FKOKy"
>
  <header
    class="sc-efBctP eZxVPd"
  >
    <a
      href="/"
    >
      <img
        alt="home logo"
        class="sc-BeQoi"
        src="logo.png"
      />
    </a>
  </header>
  <div
    class="sc-cxabCf bsgFzP"
  >
    <div
      class="sc-cxabCf iIlsBP"
    >
      <table
        class="sc-WCkqM sc-kngDgl GQgxF bjOxoH"
      >
        <thead
          class="sc-kYWVYA kzCKnj"
        >
          <tr
            class="sc-jGprRt dezIds"
          >
            <th
              class="sc-tsFYE sc-afnQL jgyunm RCrBC"
            >
              name
            </th>
            <th
              class="sc-tsFYE sc-afnQL jgyunm RCrBC"
            >
              description
            </th>
          </tr>
        </thead>
        <tbody
          class="sc-ikjQzJ"
        >
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Zazio
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Drainage of Subdural Space with Drain Dev, Open Approach
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Camido
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Supplement Right Foot Vein with Nonaut Sub, Open Approach
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Kimia
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Drainage of Bilateral Breast, Open Approach
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Pixoboo
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Bypass Bi Vas Deferens to L Vas Def w Nonaut Sub, Open
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Dabvine
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Excision of Colic Vein, Open Approach, Diagnostic
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Yakidoo
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Bypass Innom Art to L Up Arm Art w Synth Sub, Open
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Roomm
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Replace of R Metacarpophal Jt with Autol Sub, Open Approach
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Avaveo
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Plain Radiography of Bi Lacrimal Duct using L Osm Contrast
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Tagchat
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Stereotactic Particulate Radiosurgery of Ovary
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Avaveo
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Destruction of Superior Mesenteric Vein, Perc Endo Approach
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Skyvu
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Inspection of Right Upper Arm, Open Approach
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Rhynoodle
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Drainage of Right Nipple, Endo
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Voonyx
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Motor Function Treatment of Integu Up Back/UE
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Aivee
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Drainage of R Kidney Pelvis with Drain Dev, Open Approach
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Brainbox
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Transfuse Autol Fresh Plasma in Periph Art, Perc
            </td>
          </tr>
          <tr
            class="sc-jGprRt ciIpUr"
          >
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Twitternation
            </td>
            <td
              class="sc-tsFYE sc-afnQL liZNfH RCrBC"
            >
              Dilation of Tricuspid Valve with Intralum Dev, Perc Approach
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
